src/kafka_provider.py

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""KafkaProvider class and methods."""
+
+import logging
+import secrets
+import string
+from typing import Dict
+
+from ops.charm import RelationBrokenEvent, RelationJoinedEvent
+from ops.framework import Object
+from ops.model import Relation
+
+REL_NAME = "kafka"
+PEER = "cluster"
+
+logger = logging.getLogger(__name__)
+
+
+class KafkaProvider(Object):
+    """Implements the provider-side logic for client applications relating to Kafka."""
+
+    def __init__(self, charm) -> None:
+        super().__init__(charm, "client")
+
+        self.charm = charm
+
+        self.framework.observe(
+            self.charm.on[REL_NAME].relation_joined, self._on_client_relation_joined
+        )
+        self.framework.observe(
+            self.charm.on[REL_NAME].relation_broken, self._on_client_relation_broken
+        )
+
+    @property
+    def app_relation(self) -> Relation:
+        """The Kafka cluster's peer relation."""
+        return self.charm.model.get_relation(PEER)
+
+    def relation_config(self, relation: Relation) -> Dict[str, str]:
+        """Builds necessary relation data for a given relation.
+
+        Args:
+            event: the event needing config
+
+        Returns:
+            Dict of `username`, `password` and `endpoints` data for the related app
+        """
+        username = f"relation-{relation.id}"
+        password = self.app_relation.data[self.charm.app].get(username, self.generate_password())
+        units = set([self.charm.unit] + list(self.app_relation.units))
+        endpoints = [self.app_relation.data[unit]["private-address"] for unit in units]
+
+        return {"username": username, "password": password, "endpoints": ",".join(endpoints)}
+
+    def _on_client_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Handler for `relation_joined` events."""
+        if not self.charm.unit.is_leader():
+            return
+
+        relation_config = self.relation_config(relation=event.relation)
+
+        self.add_user(username=relation_config["username"], password=relation_config["password"])
+        event.relation.data[self.charm.app].update(
```

Comment:
you could just do `event.relation.data[self.charm.app].update(relation_config)` here since all the update keys are the keys in relation config